ckan_datapackage_tools/converter.py

```diff
@@ -177,82 +174,169 @@ def _parse_notes(dataset_dict):
     return result
 
 
-def _parse_license(dataset_dict):
+def _parse_tags(dataset_dict):
     result = {}
-    license = {}
 
-    if dataset_dict.get('license_id'):
-        license['type'] = dataset_dict['license_id']
-    if dataset_dict.get('license_title'):
-        license['title'] = dataset_dict['license_title']
-    if dataset_dict.get('license_url'):
-        license['url'] = dataset_dict['license_url']
+    keywords = [tag['name'] for tag in dataset_dict.get('tags', [])]
 
-    if license:
-        result['license'] = license
+    if keywords:
+        result['keywords'] = keywords
 
     return result
 
 
-def _parse_author_and_source(dataset_dict):
+def _parse_extras(dataset_dict):
     result = {}
-    source = {}
-
-    if dataset_dict.get('author'):
-        source['name'] = dataset_dict['author']
-    if dataset_dict.get('author_email'):
-        source['email'] = dataset_dict['author_email']
-    if dataset_dict.get('url'):
-        source['web'] = dataset_dict['url']
+    extras = [[extra['key'], extra['value']] for extra
+              in dataset_dict.get('extras', [])]
+    for extra in extras:
+        try:
+            extra[1] = json.loads(extra[1])
+        except (ValueError, TypeError):
+            pass
+    sources = _parse_organization_as_primary_source(dataset_dict)
+    licenses = _parse_primary_license(dataset_dict)
+    contributors = _parse_primary_contributors(dataset_dict)
+    if extras:
+        extras_dict = dict(extras)
+        result.update(_parse_profile(extras_dict))
+        _parse_extra_sources(extras_dict, sources)
+        _parse_extra_licenses(extras_dict, licenses)
```

Comment:
AFAICT this PR does not include changes to support ingesting data packages so I don't know how are you planning to implement it, but I would assume that if a DP had two licenses in the `licenses` property they would be both kept in the `licenses` extra, regardless off whether the first is used to populate `license_id` and co. If that is the case (which would be my preference), this will create a duplicated license when exporting a DP.

```json
{
  "name": "my-dp",
  "licenses": [
    {"name": "cc-zero"},
    {"name": "cc-by-sa"}
]
}
```

would generate the following CKAN dataset:

```json
{
  "name": "my-dp",
  "license_id": "cc-zero",
  "extras": [
    {"key": "licenses", "value": [
      {"name": "cc-zero"},
      {"name": "cc-by-sa"}
   ]
   }
  ]
}
```

With this code the output dp would be

```json
{
  "name": "my-dp",
  "licenses": [
    {"name": "cc-zero"},
    {"name": "cc-zero"},
    {"name": "cc-by-sa"}
]
}
```
Basically we need a check on `_parse_extra_licenses()` to see if a license already exists in the list (and the same approach should be followed by any property of which CKAN stores one item and DP allows many).

Does this make sense?

Reply:
Hi @amercader 
Yes the ingest will be following shortly...
Yes I can duplicate the licenses in extras, if that follows the overall pattern you had in mind, but I'm not sure I follow why? From comments I've seen in other parts of the code I thought the idea was that, for ingest, extras would/should (eventually) simply contain what remains after relevant properties are populated? So in the ingest that follows, the 'primary license' pops the first license from the license array?

Reply:
I know it makes sense purely from a ckanext-datapackager point of view to not duplicate the first license, but from a general integration point of view if other third party tools want to interact with the `licenses`  extra they won't necessarily know that this only contains n -1 licenses and they need to go find the other one on `license_id`

Reply:
Hi @amercader 
Yep makes sense

Reply:
Hi @amercader 
Sorry for closing this and other issue - there's been change in project requirements on our side - I'll pick this up again with you outside our project, so that I can at least provide you with updates for this and ckanext-datapackager - happy to keep working through what I have already with you until you are satisfied with the PRs

Reply:
No problem @mattRedBox thanks for letting me know